File: AkkaWorkspace/src/main/java/com/renho/akka/java/actors/MasterActor.java
package com.renho.akka.java.actors;

import akka.actor.ActorRef;
import akka.actor.Props;
import akka.actor.UntypedActor;
import com.renho.akka.java.messages.Result;

/**
 * Created by renho on 2016/3/7.
 */
public class MasterActor extends UntypedActor {

    private ActorRef aggregateActor = getContext().actorOf(Props.create(AggregateActor.class), "aggregate");
    private ActorRef reduceActor = getContext().actorOf(Props.create(ReduceActor.class, aggregateActor), "reduce");
    private ActorRef mapActor = getContext().actorOf(Props.create(MapActor.class, reduceActor), "map");

    @Override
    public void onReceive(Object message) throws Exception {
            if(message instanceof String) {
                mapActor.tell(message, this.getSelf());
            } else if(message instanceof Result) {
                aggregateActor.tell(message, this.getSelf());
            } else {
                unhandled(message);
            }
    }
}

File: akka-scala/src/main/scala/com/renho/akka/scala/actors/MasterActor.scala
package com.renho.akka.scala.actors

import akka.actor.{ActorRef, Props, Actor}
import com.renho.akka.scala.messages.{BackResult, Result}

/**
  * Created by renho on 2016/3/7.
  */
class MasterActor extends  Actor{
	val aggregateActor: ActorRef = context.actorOf(Props[AggregateActor], name="aggregate")
	val reduceActor: ActorRef = context.actorOf(Props(new ReduceActor(aggregateActor)), name="reduce")
	val mapActor: ActorRef = context.actorOf(Props(new MapActor(reduceActor)), name="map")

	println(aggregateActor.path)
	println(reduceActor.path)
	println(mapActor.path)

	override def receive: Receive = {
		case message: String =>
//			mapActor ! message
			//Forwards the message and passes the original sender actor as the sender.
			mapActor.forward(message)
		case message: Result => aggregateActor ! message
		case message: BackResult => sender() ! "back result string"
		case _ =>
	}

	override def postStop(): Unit = {
		println("PostStop");
	}
}
